{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "33910f72",
   "metadata": {},
   "source": [
    "# ENCODE bigWig analysis questions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cbbd8a6",
   "metadata": {},
   "source": [
    "Do your best to answer all parts of each question. You are encouraged to work in pairs.\n",
    "\n",
    "Answers to many questions may benefit from using more than one cell, as well as a combination of markdown and code cells.\n",
    "\n",
    "Put helper functions into a separate script (e.g. hwutils.py) so they can be commented on easily and focus the notebook on plotting. Also see the workshop on Clean Code from Dan Larremore (https://larremorelab.github.io/slides/) for good coding tips to use in this assignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8e3cedc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# useful libraries to import\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import  sklearn.decomposition\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import EngFormatter\n",
    "bp_formatter = EngFormatter('b') \n",
    "# nice way to format ticks as human-readable: ax.xaxis.set_major_formatter(bp_formatter)\n",
    "\n",
    "##hwutils is a separate script where we can put helper functions\n",
    "#from hwutils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "926d8736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/vh/c9rgmxsx7rx9w6jhh5l01ymc0000gn/T/ipykernel_8170/3550046313.py:7: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  bigwig_metadata = pd.read_table('/Volumes/GoogleDrive/My Drive/QBIO577/Data/ENCODE_GRCh38_bigWig_metadata.tsv')\n"
     ]
    }
   ],
   "source": [
    "# load dataFrame of bigWigs from ENCODE (encodeproject.org/), binned to 10kb resolution across chromosome 10.\n",
    "# note that the first three columns are chrom,start,end and the other columns are labeled by bigWig file accession.\n",
    "df = pd.read_table('/Volumes/GoogleDrive/My Drive/QBIO577/Data/ENCODE_GRCh38_binned_subset.tsv')\n",
    "\n",
    "# load metadata from ENCODE for bigwig files. \n",
    "# can be queried as follows: bigwig_metadata.query(\"`File accession`==@ df_column_name \")\n",
    "bigwig_metadata = pd.read_table('/Volumes/GoogleDrive/My Drive/QBIO577/Data/ENCODE_GRCh38_bigWig_metadata.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3da3354f",
   "metadata": {},
   "source": [
    "After loading the data (above), and visualize some of the profiles. Why might many signals dip on chr10 at around 40Mb?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "94ec77d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>File accession</th>\n",
       "      <th>File format</th>\n",
       "      <th>File type</th>\n",
       "      <th>File format type</th>\n",
       "      <th>Output type</th>\n",
       "      <th>File assembly</th>\n",
       "      <th>Experiment accession</th>\n",
       "      <th>Assay</th>\n",
       "      <th>Donor(s)</th>\n",
       "      <th>...</th>\n",
       "      <th>Genome annotation</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Controlled by</th>\n",
       "      <th>File Status</th>\n",
       "      <th>s3_uri</th>\n",
       "      <th>File analysis title</th>\n",
       "      <th>File analysis status</th>\n",
       "      <th>Audit WARNING</th>\n",
       "      <th>Audit NOT_COMPLIANT</th>\n",
       "      <th>Audit ERROR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>38</td>\n",
       "      <td>ENCFF298EKM</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>signal p-value</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR000ANO</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO156MSV/</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2017/02/07/7220c4f7-ab3a-43...</td>\n",
       "      <td>ENCODE3 GRCh38</td>\n",
       "      <td>archived</td>\n",
       "      <td>borderline replicate concordance, low read dep...</td>\n",
       "      <td>severe bottlenecking, poor library complexity</td>\n",
       "      <td>extremely low read depth, missing control alig...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39</td>\n",
       "      <td>ENCFF599PYU</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>signal p-value</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR000ANO</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO269AAA/</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2017/02/07/27c489a3-e834-45...</td>\n",
       "      <td>ENCODE3 GRCh38</td>\n",
       "      <td>archived</td>\n",
       "      <td>borderline replicate concordance, low read dep...</td>\n",
       "      <td>severe bottlenecking, poor library complexity</td>\n",
       "      <td>extremely low read depth, missing control alig...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40</td>\n",
       "      <td>ENCFF393FQJ</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>signal p-value</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR000ANO</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO156MSV/, /human-donors/ENCD...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2017/02/07/86f9d2fb-ca91-40...</td>\n",
       "      <td>ENCODE3 GRCh38</td>\n",
       "      <td>archived</td>\n",
       "      <td>borderline replicate concordance, low read dep...</td>\n",
       "      <td>severe bottlenecking, poor library complexity</td>\n",
       "      <td>extremely low read depth, missing control alig...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>41</td>\n",
       "      <td>ENCFF528ORD</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fold change over control</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR000ANO</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO269AAA/</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2017/02/07/26e29d25-4dae-47...</td>\n",
       "      <td>ENCODE3 GRCh38</td>\n",
       "      <td>archived</td>\n",
       "      <td>borderline replicate concordance, low read dep...</td>\n",
       "      <td>severe bottlenecking, poor library complexity</td>\n",
       "      <td>extremely low read depth, missing control alig...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>42</td>\n",
       "      <td>ENCFF911FWI</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fold change over control</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR000ANO</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO156MSV/, /human-donors/ENCD...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2017/02/07/450bbbc3-d0ba-45...</td>\n",
       "      <td>ENCODE3 GRCh38</td>\n",
       "      <td>archived</td>\n",
       "      <td>borderline replicate concordance, low read dep...</td>\n",
       "      <td>severe bottlenecking, poor library complexity</td>\n",
       "      <td>extremely low read depth, missing control alig...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90822</th>\n",
       "      <td>703035</td>\n",
       "      <td>ENCFF319PNN</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fold change over control</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR461MQX</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO000AAZ/</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2020/06/30/1f32734f-2d1d-4d...</td>\n",
       "      <td>ENCODE4 v1.5.0 GRCh38</td>\n",
       "      <td>released</td>\n",
       "      <td>moderate library complexity, mild to moderate ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90823</th>\n",
       "      <td>703036</td>\n",
       "      <td>ENCFF995WZQ</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fold change over control</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR461MQX</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO000AAZ/</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2020/06/30/76bbe4b4-1ea0-40...</td>\n",
       "      <td>ENCODE4 v1.5.0 GRCh38</td>\n",
       "      <td>released</td>\n",
       "      <td>moderate library complexity, mild to moderate ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90824</th>\n",
       "      <td>703041</td>\n",
       "      <td>ENCFF765CVX</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>signal p-value</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR461MQX</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO000AAZ/</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2020/06/30/e855c4c8-6342-4d...</td>\n",
       "      <td>ENCODE4 v1.5.0 GRCh38</td>\n",
       "      <td>released</td>\n",
       "      <td>moderate library complexity, mild to moderate ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90825</th>\n",
       "      <td>703042</td>\n",
       "      <td>ENCFF706QAR</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fold change over control</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR461MQX</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO000AAZ/</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2020/06/30/dae3717e-834b-4e...</td>\n",
       "      <td>ENCODE4 v1.5.0 GRCh38</td>\n",
       "      <td>released</td>\n",
       "      <td>moderate library complexity, mild to moderate ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90826</th>\n",
       "      <td>703043</td>\n",
       "      <td>ENCFF329WRM</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>bigWig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>signal p-value</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENCSR461MQX</td>\n",
       "      <td>TF ChIP-seq</td>\n",
       "      <td>/human-donors/ENCDO000AAZ/</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>released</td>\n",
       "      <td>s3://encode-public/2020/06/30/8f675080-b37a-4c...</td>\n",
       "      <td>ENCODE4 v1.5.0 GRCh38</td>\n",
       "      <td>released</td>\n",
       "      <td>moderate library complexity, mild to moderate ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90827 rows Ã— 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0 File accession File format File type  File format type  \\\n",
       "0              38    ENCFF298EKM      bigWig    bigWig               NaN   \n",
       "1              39    ENCFF599PYU      bigWig    bigWig               NaN   \n",
       "2              40    ENCFF393FQJ      bigWig    bigWig               NaN   \n",
       "3              41    ENCFF528ORD      bigWig    bigWig               NaN   \n",
       "4              42    ENCFF911FWI      bigWig    bigWig               NaN   \n",
       "...           ...            ...         ...       ...               ...   \n",
       "90822      703035    ENCFF319PNN      bigWig    bigWig               NaN   \n",
       "90823      703036    ENCFF995WZQ      bigWig    bigWig               NaN   \n",
       "90824      703041    ENCFF765CVX      bigWig    bigWig               NaN   \n",
       "90825      703042    ENCFF706QAR      bigWig    bigWig               NaN   \n",
       "90826      703043    ENCFF329WRM      bigWig    bigWig               NaN   \n",
       "\n",
       "                    Output type File assembly Experiment accession  \\\n",
       "0                signal p-value        GRCh38          ENCSR000ANO   \n",
       "1                signal p-value        GRCh38          ENCSR000ANO   \n",
       "2                signal p-value        GRCh38          ENCSR000ANO   \n",
       "3      fold change over control        GRCh38          ENCSR000ANO   \n",
       "4      fold change over control        GRCh38          ENCSR000ANO   \n",
       "...                         ...           ...                  ...   \n",
       "90822  fold change over control        GRCh38          ENCSR461MQX   \n",
       "90823  fold change over control        GRCh38          ENCSR461MQX   \n",
       "90824            signal p-value        GRCh38          ENCSR461MQX   \n",
       "90825  fold change over control        GRCh38          ENCSR461MQX   \n",
       "90826            signal p-value        GRCh38          ENCSR461MQX   \n",
       "\n",
       "             Assay                                           Donor(s)  ...  \\\n",
       "0      TF ChIP-seq                         /human-donors/ENCDO156MSV/  ...   \n",
       "1      TF ChIP-seq                         /human-donors/ENCDO269AAA/  ...   \n",
       "2      TF ChIP-seq  /human-donors/ENCDO156MSV/, /human-donors/ENCD...  ...   \n",
       "3      TF ChIP-seq                         /human-donors/ENCDO269AAA/  ...   \n",
       "4      TF ChIP-seq  /human-donors/ENCDO156MSV/, /human-donors/ENCD...  ...   \n",
       "...            ...                                                ...  ...   \n",
       "90822  TF ChIP-seq                         /human-donors/ENCDO000AAZ/  ...   \n",
       "90823  TF ChIP-seq                         /human-donors/ENCDO000AAZ/  ...   \n",
       "90824  TF ChIP-seq                         /human-donors/ENCDO000AAZ/  ...   \n",
       "90825  TF ChIP-seq                         /human-donors/ENCDO000AAZ/  ...   \n",
       "90826  TF ChIP-seq                         /human-donors/ENCDO000AAZ/  ...   \n",
       "\n",
       "      Genome annotation Platform Controlled by File Status  \\\n",
       "0                   NaN      NaN           NaN    released   \n",
       "1                   NaN      NaN           NaN    released   \n",
       "2                   NaN      NaN           NaN    released   \n",
       "3                   NaN      NaN           NaN    released   \n",
       "4                   NaN      NaN           NaN    released   \n",
       "...                 ...      ...           ...         ...   \n",
       "90822               NaN      NaN           NaN    released   \n",
       "90823               NaN      NaN           NaN    released   \n",
       "90824               NaN      NaN           NaN    released   \n",
       "90825               NaN      NaN           NaN    released   \n",
       "90826               NaN      NaN           NaN    released   \n",
       "\n",
       "                                                  s3_uri  \\\n",
       "0      s3://encode-public/2017/02/07/7220c4f7-ab3a-43...   \n",
       "1      s3://encode-public/2017/02/07/27c489a3-e834-45...   \n",
       "2      s3://encode-public/2017/02/07/86f9d2fb-ca91-40...   \n",
       "3      s3://encode-public/2017/02/07/26e29d25-4dae-47...   \n",
       "4      s3://encode-public/2017/02/07/450bbbc3-d0ba-45...   \n",
       "...                                                  ...   \n",
       "90822  s3://encode-public/2020/06/30/1f32734f-2d1d-4d...   \n",
       "90823  s3://encode-public/2020/06/30/76bbe4b4-1ea0-40...   \n",
       "90824  s3://encode-public/2020/06/30/e855c4c8-6342-4d...   \n",
       "90825  s3://encode-public/2020/06/30/dae3717e-834b-4e...   \n",
       "90826  s3://encode-public/2020/06/30/8f675080-b37a-4c...   \n",
       "\n",
       "         File analysis title File analysis status  \\\n",
       "0             ENCODE3 GRCh38             archived   \n",
       "1             ENCODE3 GRCh38             archived   \n",
       "2             ENCODE3 GRCh38             archived   \n",
       "3             ENCODE3 GRCh38             archived   \n",
       "4             ENCODE3 GRCh38             archived   \n",
       "...                      ...                  ...   \n",
       "90822  ENCODE4 v1.5.0 GRCh38             released   \n",
       "90823  ENCODE4 v1.5.0 GRCh38             released   \n",
       "90824  ENCODE4 v1.5.0 GRCh38             released   \n",
       "90825  ENCODE4 v1.5.0 GRCh38             released   \n",
       "90826  ENCODE4 v1.5.0 GRCh38             released   \n",
       "\n",
       "                                           Audit WARNING  \\\n",
       "0      borderline replicate concordance, low read dep...   \n",
       "1      borderline replicate concordance, low read dep...   \n",
       "2      borderline replicate concordance, low read dep...   \n",
       "3      borderline replicate concordance, low read dep...   \n",
       "4      borderline replicate concordance, low read dep...   \n",
       "...                                                  ...   \n",
       "90822  moderate library complexity, mild to moderate ...   \n",
       "90823  moderate library complexity, mild to moderate ...   \n",
       "90824  moderate library complexity, mild to moderate ...   \n",
       "90825  moderate library complexity, mild to moderate ...   \n",
       "90826  moderate library complexity, mild to moderate ...   \n",
       "\n",
       "                                 Audit NOT_COMPLIANT  \\\n",
       "0      severe bottlenecking, poor library complexity   \n",
       "1      severe bottlenecking, poor library complexity   \n",
       "2      severe bottlenecking, poor library complexity   \n",
       "3      severe bottlenecking, poor library complexity   \n",
       "4      severe bottlenecking, poor library complexity   \n",
       "...                                              ...   \n",
       "90822                                            NaN   \n",
       "90823                                            NaN   \n",
       "90824                                            NaN   \n",
       "90825                                            NaN   \n",
       "90826                                            NaN   \n",
       "\n",
       "                                             Audit ERROR  \n",
       "0      extremely low read depth, missing control alig...  \n",
       "1      extremely low read depth, missing control alig...  \n",
       "2      extremely low read depth, missing control alig...  \n",
       "3      extremely low read depth, missing control alig...  \n",
       "4      extremely low read depth, missing control alig...  \n",
       "...                                                  ...  \n",
       "90822                                               None  \n",
       "90823                                               None  \n",
       "90824                                               None  \n",
       "90825                                               None  \n",
       "90826                                               None  \n",
       "\n",
       "[90827 rows x 59 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df\n",
    "#bigwig_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a424f936",
   "metadata": {},
   "source": [
    "Use scikit-learn to perform PCA, and make a scatterplot of PC1 vs PC2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "533dbe80",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4d120d48",
   "metadata": {},
   "source": [
    "Try to use the experiment metadata to understand and remove outliers. Try labeling or coloring points by various metadata columns. Were any columns in the metadata useful for outlier removal? Note that sklearn.preprocessing.LabelEncoder() can be useful for transforming text strings to categories, and plt.text can be used to overlay labels with points."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7ff1383",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d52b4be2",
   "metadata": {},
   "source": [
    "Which Assays or Experiment Targets show broad vs narrow patterns? Is this consistent across cell types? Does this relate to the patterns seen in PCA? One way to investigate the characteristic scale is by computing the autocorrelation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0f36971",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "61b623e9",
   "metadata": {},
   "source": [
    "Which \"Experiment Targets\" (e.g. histone marks or transcription factors) for which cell types are nearby in this PC1 vs PC2 space? Do any of these proximities have plausible biological interpretations? For example, are any polycomb-related factors in proximity? Illustrate this in a plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73d9d16e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "eb754838",
   "metadata": {},
   "source": [
    "How much does preprocessing matter? Try normalizing the variance per track and see if you arrive at similar or distinct conclusions. Try removing the region on chr10 mentioned above. Note that sklearn.preprocessing.StandardScaler could be useful for preprocessing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "663f756e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5c255615",
   "metadata": {},
   "source": [
    "How many PCs are needed to explain 90% of the variance in the data? Illustrate this with a scree plot (https://en.wikipedia.org/wiki/Scree_plot).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54547f17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1762db04",
   "metadata": {},
   "source": [
    "How different is the dimensionality reduction into two dimensions for PCA from that obtained using MDS (multi-dimensional scaling)? What methods could be used to determine the similarity? Illustrate with a plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09feebc0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6fd2eb1e",
   "metadata": {},
   "source": [
    "Would non-negative matrix factorization (https://en.wikipedia.org/wiki/Non-negative_matrix_factorization) be a useful method to use for this dataset? Why or why not? (No plots needed for this question)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
